src/pybind/mgr/orchestrator.py

```diff
@@ -384,36 +386,158 @@ class DriveGroupSpec(object):
     """
     Describe a drive group in the same form that ceph-volume
     understands.
+    This structure allows the selection of storage devices
+    in a set of nodes
     """
-    def __init__(self, devices):
-        self.devices = devices
+    def _init_attribute(self, att_dict, attribute_name, default_value):
+        """ Extract attribute from dict, if not present applies default value
+
+        :param att_dict      : dictionary with attributes
+        :param attribute_name: The attribute to initialize
+        :param default_value : Value applied if the attribute is not in the dict
+        """
+        if attribute_name in att_dict.keys():
+            value = att_dict[attribute_name]
+        else:
+            value = default_value
+
+        setattr(self, attribute_name, value)
+
+    def _verify_device_especification(self, dev_spec):
+        """Verifies a drive group device specification
+
+        Attributes supported in the device specification are:
+            {
+            id_model : A wildcard string. e.g: "SDD*"
+            size     : Size specification of format LOW:HIGH.
+                        Can also take the the form :HIGH, LOW:
+                        or an exact value (as ceph-volume inventory reports)
+            rotates  : is the drive rotating or not
+            count    : if this is present limit the number of drives to this
+                        number.
+            }
+
+        Any attributes (even none) can be included in the device
+        specification structure.
+
+        :param device_spec: Dict with the device specification desired
+        """
+        # A dict needed
+        if type(dev_spec) is not dict:
+                raise DriveGroupError("Device specification must be a \
+                                        dictionary")
+
+        # Only certain attributes are allowed:
+        allowed_att = frozenset(["id_model", "size", "rotates", "count"])
+        att_dev_spec = frozenset(dev_spec.keys())
+        if len(att_dev_spec - allowed_att):
+            raise DriveGroupError("Unsupported attributes found \
+                                    in device specification")
+
+        return True
+
+    def __init__(self, json_drive_group):
```

Comment:
can you please call this
```python
@staticmethod
def from_json(...
```

Reply:
Your current implementation is going to prevent everyone from creating a `DriveGroupSpec` without generating a JSON representation first. Also, you're modifying a specification that is going to be used by different external orchestrators: You don't know what other orchestrators need. And finally: You're creating an exception here: no other class `__init__` implicitly expects a json structure.